### ISSUE TITLE
OS X support

### ISSUE THREAD

**Issue by kovidgoyal** [kovidgoyal/kitty#5]
Issue to track the status of porting kitty to OS X. 

Things needed to port:

0) ~~Fix any cross platform build issues when compiling the C code parts of kitty on OS X. Once it is building, setup continuous integration testing for OS X as well~~
1) ~~Replace use of xdpyinfo  to get screen DPI with native OS X API (CGDisplayBounds and CGDisplayScreenSize?)~~
2) ~~Replace use of freetype+fontconfig to rasterize fonts with native OS X APIs (this is low priority since it is possible to setup fontconfig+freetype on OSX as well)~~

Instructions for running kitty on OS X
---------------------------------------------------------

kitty now builds on OS X in Travis and all tests pass. A simple command to install the dependencies needed for kitty using brew:

```sh
brew update && brew install python3 && brew install glfw
```

Then simply follow the instructions in the README for building kitty from source.

**Comment by toomim** [comment]
Here's a record of my OSX build attempt:

    brew install glew                             # install glew
    git clone https://github.com/glfw/glfw.git    # install glfw
    cd glfw
    cmake -DBUILD_SHARED_LIBS=ON .
    make install
    cd ..
    git clone https://github.com/kovidgoyal/kitty # install kitty
    cd kitty
    # Then delete ", '-l' + lib" on line 78 of setup.py... and run:
    python setup.py install

But I get this error:

    ld: symbol(s) not found for architecture x86_64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)

I deleted the `, '-l' + lib` from setup.py because otherwise it errors with: `ld: library not found for -lPython.framework/Versions/3.5/Python`.

**Comment by kovidgoyal** [comment]
I already committed code to handle that issue, see #2

**Comment by toomim** [comment]
Ok, I'm getting further now. It turns out homebrew only has glew version 1.3, so I installed version 2 with:

```
wget https://sourceforge.net/projects/glew/files/glew/2.0.0/glew-2.0.0.tgz/download
mv download glew.tgz
tar xzvf glew.tgz
cd glew-2.0.0/
make
sudo make install
```

And now running `python3 setup.py build` produces this strange error:

```
glew >= 2.0.0 is required, found version: 2.0.0
```

**Comment by kovidgoyal** [comment]
I guess `pkg-config --atleast-version=2 glew` is not working on OS X? I suppose it could be replaced by parsing the output of pkg-config --version manually on os x.

**Comment by kovidgoyal** [comment]
Committed code to parse the version manually.

**Comment by kovidgoyal** [comment]
kitty now build on OS X in Travis. A simple script to install the dependencies needed by kitty to build:

```
#!/bin/sh
export SWBASE=/sw 
export SW=$SWBASE/sw 

sudo mkdir -p $SW && sudo chown -R $USER $SWBASE;
curl https://download.calibre-ebook.com/travis/kitty/osx.tar.xz | tar xJ -C $SW;
```

This will install the deps into the /sw (unfortunately, for technical reasons the path must be /sw). Once this is done, you can build kitty itself with:

```
PATH=$SW/bin:$PATH python3 setup.py build
```

**Comment by kovidgoyal** [comment]
Finished replacing fontconfig+freetype with OS X native APIs. This has had only the most rudimentary of testing, so YMMV

**Comment by toomim** [comment]
Cool! I got it to build and run, and crash:

```
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_GUARD
Exception Codes:       0x4000000100000004, 0x00007fff8d32d2a8

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fffa03b383a close + 10
1   org.python.python             	0x00000001058062f7 os_closerange + 87
2   org.python.python             	0x0000000105736621 _PyCFunction_FastCallDict + 497
3   org.python.python             	0x00000001057b8327 call_function + 439
4   org.python.python             	0x00000001057b4b02 _PyEval_EvalFrameDefault + 27346
5   org.python.python             	0x00000001057b958d fast_function + 381
6   org.python.python             	0x00000001057b8301 call_function + 401
7   org.python.python             	0x00000001057b4b02 _PyEval_EvalFrameDefault + 27346
8   org.python.python             	0x00000001057b8d8f _PyEval_EvalCodeWithName + 2447
9   org.python.python             	0x00000001057adf74 PyEval_EvalCode + 100
```

This looks like it's something in OSX that detects an improperly-closed file descriptor: http://stackoverflow.com/questions/32429431/exc-guard-exception

**Comment by kovidgoyal** [comment]
Hmm, closerange is used in the forked process to ensure that no file descriptors other than the stdout/strdin/stderr are open. I'm assuming that some OS X system library is opening other file descriptors. Not sure what the solution for that is.

**Comment by kovidgoyal** [comment]
I disabled closerange on OS X -- see if it helps.

**Comment by toomim** [comment]
Nope, getting the same error.

**Comment by kovidgoyal** [comment]
One more place closerange() was called, although if the code has reached there it means something else went wrong earlier. There should be something printed to stderr (in any cae I have disabled that closerange() as well.

**Comment by kovidgoyal** [comment]
kitty now builds and runs on my OS X machine -- it does not render any characters, which is likely some OpenGL bug or the other, which will need investigation.

**Comment by toomim** [comment]
It builds AND renders characters on OSX for me!

![image](https://cloud.githubusercontent.com/assets/101208/21868126/2f2e3286-d806-11e6-9262-90c833abcca7.png)

**Comment by kovidgoyal** [comment]
Glad to hear it!

I only have an ancient OS X mac mini which I keep around for building calibre -- so it's likely some OpenGL function is failing there that works on newer machines. Hopefully it works for most people.

**Comment by somebehemoth** [comment]
This is working for me on OS X 10.11.6 using your bash script.  Thank you, it looks very nice.

**Comment by kovidgoyal** [comment]
Cool, if it works for most people, I'll just forget about wasting time on figuring out the problem on my mini and just package it up into a dmg for easy installation. Fortunately, that is easy to do since I can re-use the build infrastructure I already have for calibre :) So those of you that want to run it on OS X please try it out and let me know if it works.

**Comment by shkm** [comment]
Alternaticaly / additionally, a brew (cask) formula would be great.

**Comment by kovidgoyal** [comment]
I'm not going to get into the game of creating packages for the various distro systems, on linux or OS X. However, I just committed some code to make the linux-package work on OS X as well. So now all that a creator of an OS X package has to do is (in psedo-code)

```
install dependencies: python >= 3.5, glfw >= 3.2
python3 setup.py linux-package
cp -ra linux-package/* ${installation_prefix}
```

Where installation_prefix would be whatever the package system uses (/usr/local I suppose). The run kitty with ${installation_prefix}/bin/kitty

See for example the ARCH kitty package https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=kitty-git

**Comment by wezzynl** [comment]
I also got it running on macOS Sierra (10.12.2). I run on a Retina MacBook Pro and it looks like the rendering is not Retina. Fonts look a bit blocky/fuzzy.

**Comment by kovidgoyal** [comment]
Hmm, hard to say as I cannot even get character rendering working on my ancient mac mini, let alone Retina! But, the first thing to check is if glfw is reporting the correct DPI, like this

```
kitty -c "from kitty.utils import *; from kitty.fast_data_types import *; glfw_init(); print(get_dpi())"
```

**Comment by kovidgoyal** [comment]
I can say that high dpi work fine on linux, where I use a 4K monitor (this was one of my motivations in developing kitty).

**Comment by somebehemoth** [comment]
I'm using a retina MacBook Pro (Retina, 15-inch, Mid 2015) with 10.11.6 and the text looks great.

**Comment by somebehemoth** [comment]
I am using MacBook Pro (Retina, 15-inch, Mid 2015) with 10.11.6 and the text looks nice if a bit small.  It does not seem kitty will respond to font size changes I make to kitty.conf.

**Comment by kovidgoyal** [comment]
Where is your kitty.conf located. IIRC the default location on OSX is ~/Library/Preferences/kitty/kitty.conf if it is not there then use the --config option to specify its location

**Comment by wezzynl** [comment]
Output from `kitty -c "from kitty.utils import *; from kitty.fast_data_types import *; glfw_init(); print(get_dpi())"` is:
```
{'physical': (129.30909729003906, 129.4660186767578), 'logical': (129.30909729003906, 129.4660186767578)}
```

It seems however that the issue fixes itself if I first resize the window.. The buffer inside will start to flicker and turn red. When I then maximize the window, a couple of seconds later text re-appears and then looks correct. Before the resizing, the font is also smaller and seems like it's stretched horizontally.

Not sure if this is related to the rendering issue, but I use zsh and a themed "powerline" promt which seems to screw with the rendering. I get two prompts (it renders it twice) and when I type it'll render characters twice or even three times. When I execute `bash` which has just a simple prompt, that problem disappears.

**Comment by toomim** [comment]
I can confirm that the resolution increased in the newest git. In my [old screenshot](https://github.com/kovidgoyal/kitty/issues/5#issuecomment-272008940) from git 2 days ago, each kitty pixel took up 4 of my macbook's retina pixels. 

But now, everything rendered in the window uses retina pixels. You can notice that, compared to the old screenshot, the text is now smaller in comparison to the menubar:
![image](https://cloud.githubusercontent.com/assets/101208/21947197/c3eabe24-d998-11e6-999e-a9e76510e6a9.png)

**Comment by kovidgoyal** [comment]
Yeah, this would be caused by this commit: https://github.com/kovidgoyal/kitty/commit/fa15b2d398f5fd647692753c7700c2473ed81e33

**Comment by kovidgoyal** [comment]
@shkm FYI kitty now builds against brew installed python3/glfw on Travis so it should be trivial to create a brew package for it.

**Comment by shkm** [comment]
Awesome work, @kovidgoyal! I'll take a look if I find some time and no one has done it yet.

**Comment by wezzynl** [comment]
Awesome, I agree. Thank you very much. It builds correctly against Homebrew installed deps for me on the latest Sierra.

Also. The issue I mentioned earlier about prompt not correctly rendering is fixed when I set my terminal type to `xterm-256color` instead of `xterm-kitty`.

**Comment by somebehemoth** [comment]
Just want to thank @wezzynl. I was having issues with kitty and setting TERM completely fixed them for me (zsh-syntax-highlighting, ncurses apps like tmux, ncdu, newsbeuter, and anything else that didn't like TERM set to xterm-kitty).

**Comment by kovidgoyal** [comment]
It's odd that you need to set TERM -- kitty sets the TERMINFO env var to point to the terminfo file for xterm-kitty 

Does the OS X version of terminfo not support TERMINFO env vars? On linux, zsh syntax highlighting, tig, vim, tmux etc all work fine with TERMINFO=/path/to/kitty/terminfo TERM=xterm-kitty

Maybe if you copied over the terminfo file into /usr/share/terminfo it would work without needing to set TERM.

**Comment by somebehemoth** [comment]
Setting TERM in ~/Library/Preferences/kitty/kitty.conf works as expected. I did not understand that the problems I was experiencing related to TERM setting.  This was user error.  @wezzynl 's comment reminded me to check TERM value.  iTerm2 uses "Profiles" to set TERM values so I didn't think of this setting in kitty.conf. 

Thank you for making kitty.

**Comment by kovidgoyal** [comment]
The point is you should not need to change TERM at all, because kitty comes with a terminfo file of its own. But apparently, on OS X, ncurses/terminfo does not honor the TERMINFO environment variable. So run this instead,

tic /path/to/kitty/terminfo/kitty.terminfo

that will generate the terminfo database for kitty in ~/.terminfo and then you should not need to change TERM.

**Comment by kovidgoyal** [comment]
This commit should fix the issue of the terminfo file not being picked up on OSX: https://github.com/kovidgoyal/kitty/commit/3cbb0b22dcefceab289402b168a5315e2a834078

**Comment by somebehemoth** [comment]
I can confirm this change worked as expected (thank you). Tmux, however, refuses to start unless I specify TERM to xterm-256color in kitty.conf.  I don't view this as a problem per se, just sharing the information in case it is useful to anyone.

**Comment by kovidgoyal** [comment]
It's been a long time since I used tmux, but I vaguely recall it used to have a hard-coded set of acceptable TERM values, maybe that is the issue.

**Comment by kovidgoyal** [comment]
Closing this issue since I have now made a kitty .dmg available on the releases page.
